### PR TITLE
Add ndf parser and multi blade row modules with an example

### PR DESCRIPTION
**Pull request by jjoy-work**
This Pull Request introduces ndf_parser and multi_blade_row modules and also adds an example that shows how to use the multi blade row module.
It also add log_filename_suffix argument for the launch_turbogrid function in the launcher module

### CODE REVIEW

File: examples/multi_blade_row_example.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# Copyright (c) 2023 ANSYS, Inc. All rights reserved
+"""
+.. _multi_blade_row_example:
+
+Multi blade-row meshing example
+-------------------------------
+This basic example shows how to set up a multi blade-row meshing instance and execute it in parallel
```
Comment (swilson-work) [inline examples/multi_blade_row_example.py:7]: There should not be a hyphen in "blade row".
Missing period on the sentence.

File: examples/multi_blade_row_example.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+#################################################################################
+# Create and use an MBR instance
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# Create an MBR instance, set it up for a multi-blade row case and execute.
+
+if __name__ == "__main__":
+    # Set working directory:
+    working_directory = os.getcwd()
+
+    # Set location for the NDF file
```
Comment (swilson-work) [inline examples/multi_blade_row_example.py:30]: There should not be a hard-coded path here, use something like "." instead. If absolutely necessary to use a full path (to illustrate what is needed) use a generic one e.g. "C:\Users\<name>\MBR_example".

File: examples/multi_blade_row_example.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    case_folder = r"E:\AnsysDev\pyAnsys\S947493_concepts_nrec\turb_axial_4stage_geom"
+
+    # Set name of the NDF file
+    ndf_file_name = "turb_axial_4stage_geom.ndf"
+
```
Comment (swilson-work) [inline examples/multi_blade_row_example.py:35]: Make sure all the instructions have a period at the end (or none if that is preferred, but make sure that all are consistent).

File: pyproject.toml [hunk 4]
```diff
@@ -35,6 +35,8 @@ dependencies = [
     "ansys-turbogrid-api>=0.4.dev1",
     "matplotlib>=3.5.1",
     "importlib-metadata>=4.0",
```
Comment (swilson-work) [inline pyproject.toml:37]: These dependencies are only needed to run one or more specific examples; they aren't needed for general usage of the module. I think that there should be a better way to handle these; let's discuss offline.

File: src/ansys/turbogrid/core/launcher/launcher.py [hunk 5]
```diff
@@ -148,6 +149,8 @@ def launch_turbogrid(
         an available port is automatically selected.
     host : str, default: ``127.0.0.1 (this is the local host IP for windows and linux)``
         host for TurboGrid communications. The default is ``127.0.0.1, or the local Host``
+    log_filename_suffix : str, default: ""
```
Comment (swilson-work) [inline src/ansys/turbogrid/core/launcher/launcher.py:152]: Capital "S" for suffix. Although not related to this change, the line above (151) should have a capital "H" for the first letter as well (and a small "h" in "local Host").

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py [hunk 6]
```diff
@@ -0,0 +1,1130 @@
+# Copyright (c) 2023 ANSYS, Inc. All rights reserved
```
Comment (swilson-work) [inline src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py:1]: "blade row" not "blade-row"

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py [hunk 7]
```diff
@@ -0,0 +1,1130 @@
...
+from ansys.turbogrid.api import pyturbogrid_core
+from ansys.turbogrid.api.CCL.ccl_object_db import CCLObjectDB
+from fabric import Connection
+from jinja2 import Environment, FileSystemLoader
+
+from ansys.turbogrid.core.launcher.launcher import launch_turbogrid
+from ansys.turbogrid.core.mesh_statistics import mesh_statistics
+import ansys.turbogrid.core.ndf_parser.ndf_parser as ndfp
+
+
```
Comment (swilson-work) [inline src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py:22]: I would prefer "MultiBladeRow" as the class name; it is more consistent with the naming conventions and probably easier for a user to remember.

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py [hunk 8]
```diff
@@ -0,0 +1,1130 @@
...
+        self._assert_unique_blade_names()
+        self.set_blade_rows_to_mesh([])
+
+    def set_blade_rows_to_mesh(self, blades_to_mesh: list):
+        """
+        Set the blade rows to be meshed.
+
+        Parameters
+        ----------
+        blades_to_mesh:list
```
Comment (swilson-work) [inline src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py:93]: Add spaces around the colon: "blades_to_mesh : list"

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py [hunk 9]
```diff
@@ -0,0 +1,1130 @@
...
+        self._blade_rows_to_mesh = {}
+        for blade_row in blade_row_names_to_mesh:
+            self._blade_rows_to_mesh[blade_row] = all_blade_rows[blade_row]
+        # print(f"Blade rows to mesh: {self._blade_rows_to_mesh.keys()}")
+        # for blade_row in self._blade_rows_to_mesh:
+        #    print(f"Blade row {blade_row} blades: {self._blade_rows_to_mesh[blade_row]}")
+
+    def set_multi_process_count(self, multi_process_count: int):
+        """
+        Set number of process to be used in parallel for meshing the selected blade rows.
```
Comment (swilson-work) [inline src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py:111]: Typo "processes"

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py [hunk 10]
```diff
@@ -0,0 +1,1130 @@
...
+        multi_process_count_clamped = min(max(0, multi_process_count), max_producers)
+        num_producers = min(num_rows_to_process, multi_process_count_clamped)
+        num_producers = num_producers if num_producers > 0 else max_producers
+        # print(f"{cpu_count()=}")
+        # print(f"{max_producers=}")
+        # print(f"{multi_process_count_clamped=}")
+        # print(f"{num_producers=}")
+        self._multi_process_count = num_producers
+
+    def set_blade_row_gsfs(self, assembly_gsf: float):
```
Comment (swilson-work) [inline src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py:132]: I couldn't guess what this function was doing from its name as I had no idea what gsfs stood for. Can this function just be called "set_global_size_factor"?

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py [hunk 11]
```diff
@@ -0,0 +1,1130 @@
...
+        Set the spanwise count of mesh elements for the different blade rows.
+
+        Parameters
+        ----------
+        stator_spanwise_count:int
+            The element count in the spanwise direction for stator blade rows.
+        rotor_spanwise_count:int
+            The element count in the spanwise direction for rotor blade rows.
+        rotor_blade_rows:list, default: []
+            List of main blade from each rotor blade row. If not provided,
```
Comment (swilson-work) [inline src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py:161]: "All" should be "all" as it continues the sentence.

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py [hunk 12]
```diff
@@ -0,0 +1,1130 @@
...
+
+    def set_blade_first_element_offsets(
+        self, assembly_f_el_offset: float, custom_blade_f_el_offsets: dict = None
+    ):
+        """
+        Set the boundary layer first element offset for the blades.
+
+        Parameters
+        ----------
+        assembly_f_el_offset:float
```
Comment (swilson-work) [inline src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py:176]: "assembly_f_el_offset" is obscure. Can we just call it "assembly_offset"? (since "first element" is in the function name).
Similar for the second argument.

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py [hunk 13]
```diff
@@ -0,0 +1,1130 @@
...
+        """
+        self._blade_first_element_offsets = self._get_blade_feloffs(
+            assembly_f_el_offset, custom_blade_f_el_offsets
+        )
+        # print(f"blade_first_element_offsets {self._blade_first_element_offsets}")
+
+    def set_custom_blade_settings(self, custom_blade_settings: dict):
+        """
+        Set the boundary layer first element offset for the blades.
+
```
Comment (swilson-work) [inline src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py:190]: I don't think that the description on line 189 is correct.

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py [hunk 14]
```diff
@@ -0,0 +1,1130 @@
...
+    def _read_ndf(self, ndf_file, blade_row, progress_updates_queue):
+        try:
+            start_dt = dt.now()
+            ndf_name = os.path.split(ndf_file)[1]
+            ndf_name = ndf_name.split(".")[0]
+            progress_updates_queue.put([ndf_name, f"Starting {ndf_name} ndf reader"])
+            progress_updates_queue.put([ndf_name, f"Start time: {start_dt}"])
+
+            pytg_instance = launch_turbogrid(
+                product_version="24.1.0",
```
Comment (swilson-work) [inline src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py:536]: We shouldn't hard-code the version. Why is this needed?

File: src/ansys/turbogrid/core/multi_blade_row/summary_template.html [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    }
+  </style>
+</head>
+
+<body>
+
+  <h1>TurboGrid Multi Blade Row Summary</h1>
+  
+  <p>Case: {{case_name}}</p>
+
```
Comment (swilson-work) [inline src/ansys/turbogrid/core/multi_blade_row/summary_template.html:36]: The headings should be consistent about whether they use capital letters for all words or just the first. Maybe use "Mesh Size Summary" and (lower down) "Meshing Errors".

**Comment by swilson-work** [comment]
We probably need to update the copyright year to "2024".

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py (hunk 14 above)
Comment (jjoy-work) [reply src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py:538]: I was about to remove this and then just realized that the read_tginit hotfix that was introduced recently and used by MultiBladeRow is only available since 24.1. So at least 24.1 is required. I guess there is no way to specify minimum version.

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py (hunk 12 above)
Comment (jjoy-work) [reply src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py]: Changed all "_f_el_offset" to "_bl_offset". Also changed the method name to: set_blade_boundary_layer_offsets. I guess this name will give a hint on what "bl" stands for.

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py (hunk 14 above)
Comment (Acohen-work) [reply src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py:538]: What happens if we use read_tginit with 232? It should just be sending the action directly to TG.

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py [hunk 16]
```diff
@@ -0,0 +1,1138 @@
...
+    progress_updates_queue.put(
+        [
+            blade_row + "/" + blade,
+            f"Duration: {deldt_parts[0]} hours {deldt_parts[1]} minutes {deldt_parts[2]} seconds",
+        ]
+    )
+    progress_updates_queue.put([blade_row + "/" + blade, "Done"])
+
+
+def execute_ndf_blade_row_ansys_labs(
```
Comment (Acohen-work) [inline src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py:680]: Do we want to combine the functions together? It would reduce duplicated code, and also allow to include a simple 'containerized' mode, so that we could test in github as well.

**Review by Acohen-work — APPROVED** [review]
Changes suggested, but not absolutely critical at this time

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py (hunk 14 above)
Comment (jjoy-work) [reply src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py:538]: @Acohen-work, Following error occurs when product_version="23.2.0" is used:
AttributeError: 'PyTurboGrid' object has no attribute 'read_tginit'

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py (hunk 16 above)
Comment (jjoy-work) [reply src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py:680]: There are four alternate methods here: execute_ndf_bladerow, execute_tginit_bladerow,  execute_tginit_blade_row_ansys_labs and execute_ndf_blade_row_ansys_labs. The portion mostly common to all are are already in a reused method write_mesh_report below. The rest of the portions in these methods are mainly about launching PyTurboGrid and connecting to it is in the Ansys Labs methods. Other portions of the four methods are somewhat duplicate but are just the book-keeping lines for adding messages to progress updates queue etc. I think merging these methods will add many if else blocks and make the single method very long.

File: examples/multi_blade_row_example.py [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+# Perform the required imports. It is assumed that the ``ansys-turbogrid-core``
+# package has been installed.
+
+import os
+
+from ansys.turbogrid.core.multi_blade_row.multi_blade_row import MultiBladeRow
+
+#################################################################################
+# Create and use an MultiBladeRow instance
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (swilson-work) [inline examples/multi_blade_row_example.py:23]: All the inconsistencies between "multi blade-row" and "multi blade row" have already been resolved as per the previous comments. However, here we have "multi-blade row" compared to "multi blade row" elsewhere. These should be made consistent too.

**Review by swilson-work — APPROVED** [review]

**Review by zlilek-work — APPROVED** [review]

**Review by Acohen-work — APPROVED** [review]
